src/net/tcp.rs

```diff
@@ -492,6 +494,50 @@ impl Future for TcpStreamNewState {
     }
 }
 
+/// Trait which allows to wrap a raw file descriptor into `Item`.
+///
+/// This trait is highly unsafe, use very cautiously!
+#[cfg(unix)]
+pub trait TcpStreamExt {
+    /// Type in which the raw file descriptor will be wrapped.
+    type Item;
+
+    /// Wraps the given file descriptor into the `Item` type.
+    unsafe fn from_raw_fd(fd: RawFd, handle: &Handle) -> Option<Self::Item>;
+}
+
+#[cfg(unix)]
+impl TcpStreamExt for TcpStream {
+    type Item = TcpStream;
+
+    /// Returns a new `TcpStream` from a raw fd.
+    ///
+    /// This is highly unsafe: you have to pass the fd of a *connected* and *not listening*
+    /// socket.
+    unsafe fn from_raw_fd(fd: RawFd, handle: &Handle) -> Option<TcpStream> {
+        match PollEvented::new(mio::tcp::TcpStream::from_raw_fd(fd), handle) {
+            Ok(poll) => Some(TcpStream { io: poll }),
+            _ => None,
+        }
+    }
+}
+
+#[cfg(unix)]
+impl TcpStreamExt for TcpListener {
```

Comment:
Also note that this isn't strictly necessary for TCP listeners due to the `TcpListener::from_listener` method, but I think it's fine to have it for consistency.